src/stcal/outlier_detection/median.py

```diff
@@ -0,0 +1,355 @@
+"""Compute median of large datasets in memory- and runtime-efficient ways."""
+
+from __future__ import annotations
+
+import logging
+import tempfile
+import warnings
+from pathlib import Path
+from typing import Any
+
+import numpy as np
+
+log = logging.getLogger(__name__)
+log.setLevel(logging.DEBUG)
+
+_ONE_MB = 1 << 20
+
+
+def nanmedian3D(cube: np.ndarray, overwrite_input: bool = True) -> np.ndarray:
+    """Compute the median of a cube ignoring warnings and with
+    memory efficiency optimizations. np.nanmedian always uses at least 64-bit
+    precision internally, and this is too memory-intensive. Instead, loop over
+    the median calculation to avoid the memory usage of the internal upcasting
+    and temporary array allocation. The additional runtime of this loop is
```

Comment:
If I understand the question you are asking about the runtime performance of the nanmedian3D function, right?  This conversation on the JWST PR may be relevant, although no test was done with the zeroth (time/n_groups) axis being as large as that.

When you say that it seriously slowed down runtime in ramp fitting, do you mean that a median calculation was done, or just that the entire step scaled poorly with the number of integrations?

For imaging data, where we've seen the slowest processing, the largest n_groups we have seen is only ~100, since the step takes _cal files as input instead of _calints.  However, I can look into whether there would be any coronagraphic data that had a huge number of integrations in their _calints files, as this is also used by jwst for coronagraphic data

Reply:
I tried this on a relatively large coronagraphy dataset I had lying around from [JWSTDMS-921](https://jira.stsci.edu/browse/JWSTDMS-921).  The  shape of the array going into this function was (1250, 224, 288), so still not 6000 but this is on the large side of what is processed here in practice.  The median calculation does not blow up the runtime in this case: the entire outlier detection step took only 5 seconds to run, as compared with ~1 hour for the coronagraphy-specific `align_refs` and `klip` steps.  So I don't think this is a concern for runtime but let me know if you would like to see additional tests run.